File: backend/src/validation/TradeValidator.test.ts
import { describe, it, expect } from "vitest";
import { TradeValidator } from "./TradeValidator";
import { SymbolTradingRules, ValidationErrorReason } from "../exchanges/types";

const mockBtcRules: SymbolTradingRules = {
  schemaVersion: "2.0",
  symbol: "BTCUSDT",
  exchange: "binance",
  baseAsset: "BTC",
  quoteAsset: "USDT",
  minNotional: 10.0,
  minQty: 0.00001,
  maxQty: 1000.0,
  stepSize: 0.00001,
  tickSize: 0.1,
  minPrice: 0.1,
  maxPrice: 1000000.0,
  contractSize: 1.0,
  maxLeverage: 125,
  lastUpdated: Date.now(),
};

describe("TradeValidator TypeScript Unit Tests", () => {
  it("should fail validation when metadata rules are missing", () => {
    const res = TradeValidator.validate({ symbol: "BTCUSDT", entryPrice: 50000 }, null);
    expect(res.isValid).toBe(false);
    expect(res.errorCode).toBe(ValidationErrorReason.EXCHANGE_METADATA_UNAVAILABLE);
  });

  it("should fail validation for non-positive entry price", () => {
    const res = TradeValidator.validate({ symbol: "BTCUSDT", entryPrice: -100, tradeValueUsdt: 50 }, mockBtcRules);
    expect(res.isValid).toBe(false);
    expect(res.errorCode).toBe(ValidationErrorReason.INVALID_INPUT_PARAMETERS);
  });

  it("should fail validation when order notional is below minimum notional requirement", () => {
    // $5 USDT is below minNotional of $10 USDT
    const res = TradeValidator.validate({ symbol: "BTCUSDT", entryPrice: 50000, tradeValueUsdt: 5.0 }, mockBtcRules);
    expect(res.isValid).toBe(false);
    expect(res.errorCode).toBe(ValidationErrorReason.MIN_NOTIONAL_FAILED);
    expect(res.errorMessage).toContain("below exchange minimum notional requirement of $10.00 USDT");
  });

  it("should pass validation when order notional equals or exceeds minimum notional requirement", () => {
    // $20 USDT is above minNotional of $10 USDT
    const res = TradeValidator.validate({ symbol: "BTCUSDT", entryPrice: 50000, tradeValueUsdt: 20.0 }, mockBtcRules);
    expect(res.isValid).toBe(true);
    expect(res.quantizedQuantity).toBeGreaterThan(0);
    expect(res.postRoundingNotional).toBeGreaterThanOrEqual(10.0);
  });

  it("should fail validation when quantity is below minQty", () => {
    // 0.000001 BTC is below minQty 0.00001 BTC
    const res = TradeValidator.validate({ symbol: "BTCUSDT", entryPrice: 50000, quantity: 0.000001 }, mockBtcRules);
    expect(res.isValid).toBe(false);
    expect(res.errorCode).toBe(ValidationErrorReason.MIN_QTY_FAILED);
  });

  it("should floor-quantize quantity to exact stepSize precision", () => {
    // Raw qty 0.000329 with stepSize 0.00001 -> quantizes to 0.00032 ($16 USDT > minNotional $10 USDT)
    const res = TradeValidator.validate({ symbol: "BTCUSDT", entryPrice: 50000, quantity: 0.000329 }, mockBtcRules);
    expect(res.isValid).toBe(true);
    expect(res.quantizedQuantity).toBeCloseTo(0.00032, 6);
  });

  it("should fail validation when requested leverage exceeds maxLeverage", () => {
    const res = TradeValidator.validate({ symbol: "BTCUSDT", entryPrice: 50000, tradeValueUsdt: 100, leverage: 150 }, mockBtcRules);
    expect(res.isValid).toBe(false);
    expect(res.errorCode).toBe(ValidationErrorReason.LEVERAGE_LIMIT_FAILED);
  });
});

File: mobile/app/src/test/java/com/cryptopulse/app/domain/validation/TradeValidatorTest.kt
package com.cryptopulse.app.domain.validation

import com.cryptopulse.app.domain.models.SymbolTradingRules
import org.junit.Assert.*
import org.junit.Test

class TradeValidatorTest {

    private val baseRules = SymbolTradingRules(
        symbol = "BTCUSDT",
        exchange = "Binance",
        baseAsset = "BTC",
        quoteAsset = "USDT",
        minPrice = 10000.0,
        maxPrice = 100000.0,
        tickSize = 0.01,
        minQty = 0.0001,
        maxQty = 100.0,
        stepSize = 0.0001,
        minNotional = 10.0
    )

    @Test
    fun `valid price should pass`() {
        val params = TradeValidationParams(symbol = "BTCUSDT", entryPriceStr = "50000.00", currentMarketPrice = 50000.0)
        val result = TradeValidator.validate(params, baseRules)
        assertTrue(result.errorMessage, result.isValid)
    }

    @Test
    fun `price violating tick size should fail`() {
        // Tick size is 0.01, so 50000.005 should fail
        val params = TradeValidationParams(symbol = "BTCUSDT", entryPriceStr = "50000.005", currentMarketPrice = 50000.0)
        val result = TradeValidator.validate(params, baseRules)
        assertFalse("Price with too many decimals should fail", result.isValid)
        assertEquals(ValidationErrorReason.INVALID_TICK_SIZE, result.errorCode)
    }

    @Test
    fun `price below minimum should fail`() {
        val params = TradeValidationParams(symbol = "BTCUSDT", entryPriceStr = "5000.0", currentMarketPrice = 5000.0)
        val result = TradeValidator.validate(params, baseRules)
        assertFalse("Price below minPrice should fail", result.isValid)
        assertEquals(ValidationErrorReason.PRICE_BELOW_MINIMUM, result.errorCode)
    }

    @Test
    fun `price above maximum should fail`() {
        val params = TradeValidationParams(symbol = "BTCUSDT", entryPriceStr = "150000.0", currentMarketPrice = 150000.0)
        val result = TradeValidator.validate(params, baseRules)
        assertFalse("Price above maxPrice should fail", result.isValid)
        assertEquals(ValidationErrorReason.PRICE_ABOVE_MAXIMUM, result.errorCode)
    }

    @Test
    fun `missing rules should fail with clear error`() {
        val params = TradeValidationParams(symbol = "BTCUSDT", entryPriceStr = "50000.0", currentMarketPrice = 50000.0)
        val result = TradeValidator.validate(params, null)
        assertFalse(result.isValid)
        assertEquals(ValidationErrorReason.EXCHANGE_METADATA_UNAVAILABLE, result.errorCode)
    }

    @Test
    fun `very small tick size handling`() {
        val rules = baseRules.copy(tickSize = 0.00000001, minPrice = 0.0, maxPrice = 1.0)
        val params = TradeValidationParams(symbol = "SHIBUSDT", entryPriceStr = "0.00001234", currentMarketPrice = 0.00001234)
        val result = TradeValidator.validate(params, rules)
        assertTrue(result.isValid)
        
        val invalidParams = TradeValidationParams(symbol = "SHIBUSDT", entryPriceStr = "0.000012345", currentMarketPrice = 0.00001234)
        val invalidResult = TradeValidator.validate(invalidParams, rules)
        assertFalse("Price violating 8-decimal tick size should fail", invalidResult.isValid)
    }
    
    @Test
    fun `independent target entry price should pass when exchange bounds and tick size are valid`() {
        // Market price is 0.13, target entry price is 2.00
        val rules = baseRules.copy(minPrice = 0.01, maxPrice = 1000.0, tickSize = 0.01)
        val params = TradeValidationParams(symbol = "DOGEUSDT", entryPriceStr = "2.00", currentMarketPrice = 0.13)
        val result = TradeValidator.validate(params, rules)
        assertTrue("Independent target price (2.00 on 0.13 market) within bounds should pass", result.isValid)
    }
}
